### PR TITLE
Add violation logic for multi level tags

### PR DESCRIPTION
**Pull request by cead22**
### Fixed Issues

$ https://github.com/Expensify/App/issues/37117
PROPOSAL:

### Tests
- In old dot, in the workspace editor > tags tab, switch on the Use multiple levels of tags toggle
- Click import from spreadsheet, and upload the file in this issue description
- On new dot, tap big plus sign > request money
- Select Manual tab (from Manual | Scan | Distance)
- Enter $15 as the amount amount and tap next
- Tap the workspace from the pre-testing steps (tap the row, not the split button in the row)
- Tap the merchant row and enter a merchant
- Tap the Request $15 button
- Confirm the report preview has RBR
- Tap the report preview
- Confirm the money request preview has RBR and shows “Review Required”
- Tap the money request preview
    - Verify you see Missing Region, Missing Project, and Missing Department violations
- Tap the Region menu item, and select a region
    - Verify that the Missing Region violation disappears right away
- Repeat for Department and Project
- Tap the Region menu item, and unselect the selected region
    - Verify that the Missing Region violation appears right away
- 🛜 Switch to new dot and go offline
- Repeat steps 1-16. The report preview, money request preview should be grayed out
- Go online, and confirm everything looks the same, except nothing is grayed out
- Click on the link in the transaction thread header to go back to the expense report
    - Verify the money request preview says Missing Region next to the RBR

- [x] Verify that no errors appear in the JS console

### Offline tests
Covered in tests

### QA Steps
Same as tests

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

>[!Note]
>I posted about console warnings here https://expensify.slack.com/archives/C01GTK53T8Q/p1709336500920559

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/165133/d920afd8-cc77-4d00-9b20-0fd0d761a03f
</details>

<details>
<summary>Android: mWeb Chrome</summary>

I'm having issues running the app on android web, so will rely on reviewer testing this platform

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/165133/dd4f8b37-d21f-43ae-ae11-e370337bd6b3

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

Online

https://github.com/Expensify/App/assets/165133/363e8b52-b5c9-4b36-b8b0-1ad9f3cb02b3 

Offline

https://github.com/Expensify/App/assets/165133/363e8b52-b5c9-4b36-b8b0-1ad9f3cb02b3 x

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/165133/7292c892-7438-410b-aadf-733d89649cff

</details>

### CODE REVIEW

**Comment by cead22** [comment]
cc @yuwenmemon @allroundexperts since you reviewed https://github.com/Expensify/App/pull/34983

**Comment by cead22** [comment]
Sorry for the back and forth @parasharrajat , I thought I had linked to this PR in this issue https://github.com/Expensify/App/issues/36441, which you had asked to be a reviewer on, but I had linked a different PR

**Comment by DylanDylann** [comment]
@cead22 I was assigned to this PR by Melvin. Should I review this PR?

**Comment by allroundexperts** [comment]
I think it would be better if I review this since I have good context on this.

**Comment by cead22** [comment]
That works, thanks @allroundexperts

**Comment by allroundexperts** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by allroundexperts** [comment]
@cead22 The error for the `Project` field only shows up if there is no other error in any of the fields. I would expect it to show up always. Is this a bug?

https://github.com/Expensify/App/assets/30054992/598bbc4b-5c39-40f4-8017-4e23963d84c0

**Comment by cead22** [comment]
> I would expect it to show up always. Is this a bug?

That's not a bug in the sense that it's working as currently designed (assuming Project 3 is enabled, and "Project no longer valid" is accurate), but I had the same doubt, and had to confirm this was by design by looking at the code. That said, we may change this so that both violations appear when they both apply

The reason this is happening is because we only return 1 tag violation per request, so if we return `someTagLevelsRequired` we won't return `tagOutOfPOlicy` for the same request, and that's why when you fix one, the other appears.

File: src/libs/Violations/ViolationsUtils.ts [hunk 1]
```diff
@@ -78,6 +79,57 @@ const ViolationsUtils = {
                 if (!hasMissingTagViolation && !updatedTransaction.tag && policyRequiresTags) {
                     newTransactionViolations.push({name: CONST.VIOLATIONS.MISSING_TAG, type: 'violation'});
                 }
+            } else {
+                newTransactionViolations = reject(newTransactionViolations, {
+                    name: CONST.VIOLATIONS.SOME_TAG_LEVELS_REQUIRED,
+                });
+                newTransactionViolations = reject(newTransactionViolations, {
+                    name: CONST.VIOLATIONS.TAG_OUT_OF_POLICY,
+                });
```
Comment (allroundexperts) [inline src/libs/Violations/ViolationsUtils.ts:88]: Wouldn't `filter` be more appropriate here instead of rejecting twice?

File: src/hooks/useViolations.ts [hunk 2]
```diff
@@ -58,7 +58,34 @@ function useViolations(violations: TransactionViolation[]) {
...
     }, [violations]);
-    const getViolationsForField = useCallback((field: ViolationField) => violationsByField.get(field) ?? [], [violationsByField]);
+
+    const getViolationsForField = useCallback(
+        (field: ViolationField, data?: TransactionViolation['data']) => {
+            const currentViolations = violationsByField.get(field) ?? [];
+
+            // someTagLevelsRequired has special logic becase data.errorIndexes is a bit unique in how it denotes the tag list that has the violation
+            // tagListIndex can be 0 so we compare with undefined
+            if (currentViolations[0]?.name === 'someTagLevelsRequired' && data?.tagListIndex !== undefined && Array.isArray(currentViolations[0]?.data?.errorIndexes)) {
```
Comment (allroundexperts) [inline src/hooks/useViolations.ts:68]: Will the `currentViolations` always have a single element?

Comment (cead22) [reply src/hooks/useViolations.ts:68]: Yes, because we only return 1 violation per field per request

**Comment by cead22** [comment]
Updated

File: src/libs/Violations/ViolationsUtils.ts [hunk 3]
```diff
@@ -78,6 +79,53 @@ const ViolationsUtils = {
...
+                                },
+                            });
+                            hasInvalidTag = true;
+                            break;
+                        }
+                    }
+                    if (!hasInvalidTag) {
+                        newTransactionViolations = reject(newTransactionViolations, {
+                            name: CONST.VIOLATIONS.TAG_OUT_OF_POLICY,
+                        });
```
Comment (allroundexperts) [inline src/libs/Violations/ViolationsUtils.ts:126]: `filter` can be used here as well.

Comment (cead22) [reply src/libs/Violations/ViolationsUtils.ts:126]: There's no difference here since we're only calling reject once, so it should go over the array only once, right?

Comment (allroundexperts) [reply src/libs/Violations/ViolationsUtils.ts:126]: Yes, but aren't we preferring native JS methods for filtering after TS migration?

**Review by allroundexperts — APPROVED** [review]
Looks good other than a NAB change that is upon the engineers discretion to fix.

File: src/libs/Violations/ViolationsUtils.ts [hunk 4]
```diff
@@ -78,6 +79,53 @@ const ViolationsUtils = {
...
+                // Otherwise, we put TAG_OUT_OF_POLICY in Onyx (when applicable)
+                const errorIndexes = [];
+                for (let i = 0; i < policyTagKeys.length; i++) {
+                    const isTagRequired = policyTagList[policyTagKeys[i]].required ?? true;
+                    const isTagSelected = Boolean(selectedTags[i]);
+                    if (isTagRequired && (!isTagSelected || (selectedTags.length === 1 && selectedTags[0] === ''))) {
+                        errorIndexes.push(i);
+                    }
+                }
+                if (errorIndexes.length !== 0) {
```
Comment (blimpich) [inline src/libs/Violations/ViolationsUtils.ts:97]: NAB: since 0 is falsy in JS this can be simplified.
```suggestion
                if (errorIndexes.length) {
```

File: tests/unit/ViolationUtilsTest.js [hunk 5]
```diff
@@ -201,4 +196,77 @@ describe('getViolationsOnyxData', () => {
...
+                    tags: {
+                        Project1: {
+                            name: 'Project1',
+                            enabled: true,
+                        },
+                    },
+                    required: true,
+                },
+            };
+        });
```
Comment (blimpich) [inline tests/unit/ViolationUtilsTest.js:233]: NAB: can we add a line break in between tests?

File: tests/unit/ViolationUtilsTest.js [hunk 6]
```diff
@@ -201,4 +196,77 @@ describe('getViolationsOnyxData', () => {
...
+            someTagLevelsRequiredViolation.data = {errorIndexes: [1]};
+            result = ViolationsUtils.getViolationsOnyxData(transaction, transactionViolations, policyRequiresTags, policyTags, policyRequiresCategories, policyCategories);
+            expect(result.value).toEqual([someTagLevelsRequiredViolation]);
+
+            // Test case where transaction has all tags
+            transaction.tag = 'Accounting:Africa:Project1';
+            result = ViolationsUtils.getViolationsOnyxData(transaction, transactionViolations, policyRequiresTags, policyTags, policyRequiresCategories, policyCategories);
+            expect(result.value).toEqual([]);
+        });
+        it('should return tagOutOfPolicy when a tag is not enabled in the policy but is set in the transaction', () => {
```
Comment (blimpich) [inline tests/unit/ViolationUtilsTest.js:264]: NAB: can we add a line break in between tests?

File: tests/unit/ViolationUtilsTest.js [hunk 7]
```diff
@@ -201,4 +196,77 @@ describe('getViolationsOnyxData', () => {
             expect(result.value).not.toContainEqual([missingTagViolation]);
         });
     });
+    describe('policy has multi level tags', () => {
```
Comment (blimpich) [inline tests/unit/ViolationUtilsTest.js:199]: NAB: can we add a line break in between describe statements?

File: src/libs/Violations/ViolationsUtils.ts [hunk 8]
```diff
@@ -78,6 +79,53 @@ const ViolationsUtils = {
...
+                                name: CONST.VIOLATIONS.TAG_OUT_OF_POLICY,
+                                type: 'violation',
+                                data: {
+                                    tagName: policyTagKeys[i],
+                                },
+                            });
+                            hasInvalidTag = true;
+                            break;
+                        }
+                    }
```
Comment (blimpich) [inline src/libs/Violations/ViolationsUtils.ts:122]: NAB Can we add some line breaks to the new code for readability?

File: src/libs/Violations/ViolationsUtils.ts [hunk 9]
```diff
@@ -78,6 +79,53 @@ const ViolationsUtils = {
                 if (!hasMissingTagViolation && !updatedTransaction.tag && policyRequiresTags) {
                     newTransactionViolations.push({name: CONST.VIOLATIONS.MISSING_TAG, type: 'violation'});
                 }
+            } else {
```
Comment (blimpich) [inline src/libs/Violations/ViolationsUtils.ts:82]: NAB: can we consider extracting some part of this new code into a helper function? I find the logic hard to follow with all the if/else/for statements nested together. Abstracting out some of the logic would make this a lot more readable IMO.

**Review by blimpich — DISMISSED** [review]
NAB comments, otherwise looks good 👍

**Comment by cead22** [comment]
Updated! Please re-review. @blimpich I'm taking the NAB that I didn't reply to as non-blockers
- I kinda like the explicitness of `errorIndexes.length !== 0`
- Our style guide says not to add blank lines to separate code, and only add blank lines that are going to be followed by comments. Is searched usages of `it` and `describe` and it looks like we have a combination of having blank lines and no blank lines

**Comment by blimpich** [comment]
Sounds good! @cead22 can you help me find that style guide? I can't find one that mentions that we prefer to not use newlines to separate code?

**Review by blimpich — APPROVED** [review]

**Comment by cead22** [comment]
> Yes, but aren't we preferring native JS methods for filtering after TS migration?

I meant to address this but forgot 😬

**Comment by cead22** [comment]
I thought it was in our general style guides, but it looks like it's on our php style guide https://github.com/Expensify/Style-Guides/blob/4a2750183223b1f84398e0c1c633fa95a7512841/php.md#blank-lines

>Unnecessary blank lines without comments should be avoided

File: src/libs/Violations/ViolationsUtils.ts [hunk 10]
```diff
@@ -7,6 +7,106 @@ import type {TranslationPaths} from '@src/languages/types';
...
+    if (!hasMissingTagViolation && !updatedTransaction.tag && policyRequiresTags) {
+        newTransactionViolations.push({name: CONST.VIOLATIONS.MISSING_TAG, type: 'violation'});
+    }
+    return newTransactionViolations;
+}
+
+/**
+ * Calculates some tag levels required and missing tag violations for the given transaction
+ */
+function getTagViolationsForMultiLevelTags(
```
[the remark this answers is not in the document]
Comment (cead22) [reply src/libs/Violations/ViolationsUtils.ts:52]: Thanks! This wasn't really missing, as much as it just hadn't been implemented yet. When we added support for multi level tags, we broke the violations for tags, and this PR was meant to fix that.

We then added support for dependent tags, and then implemented `getTagViolationsForDependentTags ` and `getTagViolationForIndependentTags` in https://github.com/Expensify/App/pull/40741/files

File: src/libs/Violations/ViolationsUtils.ts [hunk 11]
```diff
@@ -7,6 +7,106 @@ import type {TranslationPaths} from '@src/languages/types';
...
+    updatedTransaction: Transaction,
+    transactionViolations: TransactionViolation[],
+    policyRequiresTags: boolean,
+    policyTagList: PolicyTagList,
+): TransactionViolation[] {
+    const policyTagKeys = Object.keys(policyTagList);
+    const selectedTags = updatedTransaction.tag?.split(CONST.COLON) ?? [];
```
Comment (allroundexperts) [inline src/libs/Violations/ViolationsUtils.ts:59]: We should have split the tags using `getTagArrayFromName` method instead. The `split` function fails if a tag contains more than one `:`. This caused https://github.com/Expensify/App/issues/50105
